### DISCUSSION TITLE
Shopify Scripts don't have access to the Customer for carts created with Storefront API

### DISCUSSION THREAD

**Discussion by duailibe** [Shopify/storefront-api-feedback#133] [Help] [answered]
We have a Script that changes payment providers based on the user that is buying but [here](https://help.shopify.com/en/manual/checkout-settings/script-editor/shopify-scripts#cart) it says that this is only available for carts created through Online Store channel.

This is unfortunate as this information is necessary for us to launch the headless app. 

Is that something that can be fixed? Is there anything we can do on our side?

**Comment by johncraigcole** [comment, marked as answer]
Hey @duailibe the scripts will run for a cart created with the Storefront API when at checkout. Recently, we also added support for line item scripts to run on the cart object itself.

**Reply by duailibe** [reply]
@johncraigcole Thanks for the response!

However it seems that the Ruby scripts don't have access to either the `Input.cart` or `Input.cart.customer` as it fails to run today only if the cart was created with the Storefront API.

**Reply by johncraigcole** [reply]
It fails in our web checkout or on the Storefront API cart object?

**Reply by duailibe** [reply]
It fails in the web checkout.

We tested like this:

- in Liquid site, add products to cart, navigate to checkout, some payment methods are hidden
- in headless site, add products to cart (with Storefront API), navigate to checkout (using the checkout url of the Storefront's cart object), all payment methods are shown

**Reply by johncraigcole** [reply]
Just tried a couple of examples hiding payment methods in a payment script based on both product on the cart, and customer tags without issue. I'm assuming on the script you have the `Online Store and Storefront API channels` selected. 

Have you tried un-publishing and re-publishing the script? If the app being used for the Storefront API request was created after the last update of the script, the script would need to be republished. I also just wanted to confirm the app used to generate the Storefront API token was from a custom/private app created in the Shopify admin.

**Reply by duailibe** [reply]
> If the app being used for the Storefront API request was created after the last update of the script, the script would need to be republished

Oh, this is definitely the case (the app was created later than the last update) and I did not try to re-publish it. I'll test it ASAP and report back.

> I also just wanted to confirm the app used to generate the Storefront API token was from a custom/private app created in the Shopify admin.

Yes, that is the case as this is a Hydrogen app.

**Reply by duailibe** [reply]
Just re-published it and it worked! You won't believe how many hours I've spent trying to work around it, but never thought about republishing it.

Thank you very much for the help!

**Reply by johncraigcole** [reply]
That's great to hear. Thinking about our documentation and how we can do better to help our developer community with the publication trick. At a minimum, I was thinking about adding a bit about this to the `Supported Channels` section of this [doc](https://help.shopify.com/en/manual/checkout-settings/script-editor/create). Any suggestions you have where we could add this context to save the next dev from this pain point?

**Reply by duailibe** [reply]
@johncraigcole Yes, that would be a valuable addition. A callout describing this caveat would be helpful because I can see others adding new apps and expecting their existing scripts to continue to work.
